### PR TITLE
List view: Allow selected block to override roving tabindex

### LINKED ISSUE #48337: TreeGrid RovingTabIndex doesn't allow tabIndex to be changed from outside the component
## The problem

While working on improvements to the List View accessibility in https://github.com/WordPress/gutenberg/pull/45135, an issue was encountered surrounding odd focusing behaviour in the List View. The List View incorporates roving tab index behaviour (via TreeGrid) that sets the taxIndex to the last focused element. This way, if a user tabs between the List View and other elements of the editor, they will find their way back to the element they were last working on.

However, if a user navigates to the editor canvas and changes which block is currently selected, when they tab back to the List View they will likely expect to land on the currently selected block within the List View. With the current behaviour, this doesn't happen. Instead of the currently selected block receiving focus, the tabIndex is still set on the last block within the list view that had focus.

## Desired fix

To fix this, it would be good for the roving tab index within the tree grid to be able to be controlled or updated in some way from outside of the component. For example, if the user changes the selection within the editor canvas, then the tabIndex within the list view should probably be updated so that the currently selected block is the one with tabIndex set. This might require refactoring how the roving tab index works.

Kudos @alexstine for figuring out the cause of the issue.

### PR DESCRIPTION
**Pull request by alexstine**
## What?

Fixes #48337

## Why?

See above.

## How?

If a block is selected, I replace the roving tabindex value with 0. Roving tabindex only keeps track of the last focused block and has no way to know about blocks selected in the canvas.

## Testing Instructions

1. Open a post or page.
2. Insert some blocks, at least three.
3. Select the second block in the canvas.
4. Open the list view and notice that the second block is selected and focus is placed on it. This is because the `tabindex="0"` is set.
5. Select the third block in the canvas.
6. Press the shortcut to open the list view. On Windows, it is alt+shift+o.
7. Notice how focus is now placed on the third block.

### Testing Instructions for Keyboard

No difference.

## Notes

I know this is not a perfect solution due to the way things could get odd with the options button. In any case, the block will always receive focus but the correct options button will not. This is because the block will have 0 tabindex which would make it the first tabbable. You can better understand this by looking at `useFocusFirstElement` hook. I think this is a fair trade-off for the benefit of not having a large refactor. The other option is, I disable the roving tabindex provider with a custom parameter and handle this in list view that way being in block-editor, I can have awareness of selected blocks.

### CODE REVIEW

**Comment by alexstine** [comment]
I am shocked this did not break E2E tests. I guess we have no tests around tabindex and what could have focus at any given time. Interesting.

**Review by andrewserong — COMMENTED** [review]
This is a clever fix, nice idea @alexstine! It's testing well for me so far, and I like the idea that we're adding a little logic to the List View component while using the roving tabIndex as a fallback. So essentially, it's simply adding additional `taxIndex` values to selected blocks to ensure they're available. I think that makes for a nice workaround.

For multi-block selections, though, I notice that there are now many blocks in the list view that can receive focus. For example, if I create 10 paragraph blocks and shift select them, and then open the the list view, tabbing will tab between each of the selected blocks in the list view, making it take longer if I wish to tab outside of the list view.

Is that an issue? If so, one idea might be to only override the tabIndex for either the first block in the selection, or the last block in the selection, whichever one seems more appropriate. If it isn't an issue, then this PR looks in good shape to me.

**Comment by alexstine** [comment]
@andrewserong So, I actually thought I should open a new issue about this. The terms in code confuse the heck out of me in Gutenberg. Selected can mean two things.
1. If the block is selected by holding down shift and pressing arrow keys.
2. If the block is currently being edited.

I really wish we as devs would stop  conflating the two because it makes it impossible to understand in what context `isSelected` actually does. I think the tabindex should not be modified if multi-selection is happening.

Any idea on how to fix it?

Thanks.

**Comment by andrewserong** [comment]
That's a great point about distinguishing between the current block that's being edited or has focus, versus being selected. You're right, I think the two concepts are conflated in API naming.

In terms of preventing the change when a multi selection is happening, we could perhaps try something like the following:

```
	const isOnlySelectedBlock = !! (
		isSelected && selectedClientIds.length === 1
	);
```

Then, in the checks where you're currently using `!! isSelected`, these would become `isOnlySelectedBlock` instead. And we wouldn't need the double negation anymore since it'll only happen once when `isOnlySelectedBlock` is defined.

Would that work?

**Comment by alexstine** [comment]
@andrewserong Yes, it seems to work. I tried to give the variable a reasonable name that describes what we are after. There is a related PR where the list view E2E tests are currently being migrated. I want to write a test for this but need to wait until it is merged. I already held it up once.

File: packages/block-editor/src/components/list-view/block.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/block.js]: There's tiny typo here `lenth` instead of `length`.

```suggestion
		isSelected && selectedClientIds.length === 1;
```

**Comment by alexstine** [comment]
@andrewserong Nice catch. Not sure why ESLint did not catch that. Not sure why the page did not error out...

**Review by andrewserong — APPROVED** [review]
I think `lenth` didn't throw an error because `selectedClientIds` is technically an object (albeit an array), so possibly the typo meant that ESLint thought the code was seeing if there was a `lenth` property. If we attempted to do anything other than a comparison check, it might have thrown an error. No matter now, though! 🙂

Thanks for the follow-ups, this is testing well for me! I agree, I think adding an e2e test could wait until the refactored tests land. LGTM!

File: test/e2e/specs/editor/various/list-view.spec.js
Comment (alexstine) [inline test/e2e/specs/editor/various/list-view.spec.js:45]: Had to remove this. ESLint would not allow me to commit.

**Review by alexstine — COMMENTED** [review]
One small note here.

File: test/e2e/specs/editor/various/list-view.spec.js
Comment (alexstine) [reply test/e2e/specs/editor/various/list-view.spec.js:45]: CC @kevin940726

Comment (kevin940726) [reply test/e2e/specs/editor/various/list-view.spec.js:45]: `geyByRole` doesn't return a promise so this is the right fix 👍 .

**Comment by alexstine** [comment]
Lets give this a try and see if anyone reports bugs. 👍

**Pull request merged**
